src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -538,23 +555,54 @@ repl_req_ptr_t RaftReplDev::applier_create_req(repl_key const& rkey, journal_typ
     }
 
     // rreq->init will allocate the block if it has linked data.
-    auto status = rreq->init(rkey, code, false /* is_proposer */, user_header, key, data_size, m_listener);
-    if (!rreq->has_linked_data()) { return rreq; }
-#ifdef _PRERELEASE
-    if (is_data_channel) {
-        if (iomgr_flip::instance()->test_flip("fake_reject_append_data_channel")) {
-            LOGINFO("Data Channel: Reject append_entries flip is triggered for rkey={}", rkey.to_string());
-            status = ReplServiceError::NO_SPACE_LEFT;
-        }
-    } else {
-        if (iomgr_flip::instance()->test_flip("fake_reject_append_raft_channel")) {
-            LOGINFO("Raft Channel: Reject append_entries flip is triggered for rkey={}", rkey.to_string());
-            status = ReplServiceError::NO_SPACE_LEFT;
-        }
-    }
-#endif
+    auto status = init_req_ctx(rreq, rkey, code, false /* is_proposer */, user_header, key, data_size, m_listener);
+
     if (status != ReplServiceError::OK) {
-        RD_LOGD("For Repl_key=[{}] alloc hints returned error={}, failing this req", rkey.to_string(), status);
+        if (status == ReplServiceError::NO_SPACE_LEFT && !is_data_channel && !rreq->is_proposer()) {
+            RD_LOGD("For Repl_key=[{}] alloc hints returned error={} , lsn={}", rkey.to_string(), status, lsn);
+
+            // only when the follower got no_space_left error in raft channel, we try to handle it. if it is a leader,
+            // we will not handle it and return fail to client.
+
+            pause_statemachine();
```

Comment:
I fail to follow the reason to pause state machine in this error case. If it is to block I/O, we should populate the no space error to HomeStore consumer, in this case, the HomeObject and pause I/O from there. Kindly be noted, some use cases it is just a general no space error, reflecting the fact of the storage engine (HomeStore), how consumer wants to act on this error, the logic can happen on the consumer. In NuBlox case, fail on no space left is Okay, because there is no shard layout and between replicas, they don't need to be on same layout. it is a back pressure to HomeBlocks layer and populate all the way to client to fail the I/O.